TEDIT: Align paralooks with charlooks, recompile all files, miscellaneous updates

rmkaplan opened the pull request:
For example, select a word in the middle of a line, extend that with right-click.  Then move (ctrl-shift) that line to the position of the caret (which is at the end of the BPD).

rmkaplan commented:
Originally this just had a few files for fixing the interaction of move with blue pending delete.  But it was hanging unapproved for a few weeks, there were mostly innocuous changes in Tedit files that overlapped with the few files originally here, and I didn't want to deal with the ensuing merge problems.  So I just pushed all of the other changed files into this PR.

Apart from the move issue, this provides the Tedit side of the line-up/line-down arrow capability and addresses the deadlock issue in  #1972 .

The changes in the most of the files had to do with renaming the fmtspec datatype to paralooks, to align the concept and code with the charlooks implementation.  This also moves the partially implmented character and paragraph styles from TEDIT-LOOKS to a new file TEDIT-STYLES.  That should make it easier (but not easy) to understand what John was aiming for.

FYI, here are a few still-open issues that I believe have been dealt with, here or in earlier PRs.   #1952  #1950 #1889 #1787  #1609 #1475  #1297  #1157   #1061

pamoroso commented:
This PR seems to be working fine on Linux Mint 22 Cinnamon and I'm not noticing anything broken or unusual.

pamoroso commented:
I'm playing a bit with the new character looks menu and it works as expected except for a known [text truncation issue](https://github.com/Interlisp/medley/issues/1962#issuecomment-2585851531).

rmkaplan commented:
I committed a very small change that should give the expected behavior with right arrow and the end-of-file

pamoroso commented:
I built the latest head of this branch but the arrow keys still don't work in TEdit, i.e. they insert black boxes. If I need some other code on a different branch I'm not sure how to integrate it with this branch.

rmkaplan commented:
You need the LLKEY PR too.

> On Feb 15, 2025, at 9:10 AM, Paolo Amoroso ***@***.***> wrote:
> 
> 
> pamoroso
>  left a comment 
> (Interlisp/medley#2021)
> I built the latest head of this branch but the arrow keys still don't work in TEdit, i.e. they insert black boxes. If I need some other code on a different branch I'm not sure how to integrate it with this branch.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.
> 
>  <https://github.com/Interlisp/medley/pull/2021#issuecomment-2661007450> <https://github.com/notifications/unsubscribe-auth/AQSTUJMF7FJK5FEOSES5UDT2P5YHVAVCNFSM6AAAAABWVDF3RWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRRGAYDONBVGA>
> 
> pamoroso
>  left a comment 
> (Interlisp/medley#2021)
>  <https://github.com/Interlisp/medley/pull/2021#issuecomment-2661007450>
> I built the latest head of this branch but the arrow keys still don't work in TEdit, i.e. they insert black boxes. If I need some other code on a different branch I'm not sure how to integrate it with this branch.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/2021#issuecomment-2661007450>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJMF7FJK5FEOSES5UDT2P5YHVAVCNFSM6AAAAABWVDF3RWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRRGAYDONBVGA>.
> You are receiving this because you authored the thread.
>

pamoroso commented:
I'm doing something wrong with Git but I don't know what.

From branch `rmk65--Tedit-move-with-overlapping-bluependingdelete` I executed `gh pr checkout 2033` of #2033 and `git pull` but the arrow keys still don't work in TEdit. If I rebuild Medley from `rmk68--Add-arrow-keys-to-LLKEY` how can I integrate the changes of the other branch `rmk65--Tedit-move-with-overlapping-bluependingdelete`?

nbriggs commented:
Check out rmk65, build your full.sysout (because there are a lot of changed files in this branch, so it's easier to do it that way), then checkout rmk68 and (LOAD 'LLKEY.LCOM) into a running full.sysout.  You don't always have to build a sysout in order to load changes.

pamoroso commented:
I followed that procedure but when loading LLKEY.COM from rmk68 Medley becomes unresponsive.

rmkaplan commented:
LLKEY may be a file that you can't load on top.  I tried also with the same result.



> On Feb 15, 2025, at 10:55 AM, Paolo Amoroso ***@***.***> wrote:
> 
> 
> pamoroso
>  left a comment 
> (Interlisp/medley#2021)
> I followed that procedure but when loading LLKEY.COM from rmk68 Medley becomes unresponsive.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.
> 
>  <https://github.com/Interlisp/medley/pull/2021#issuecomment-2661045839> <https://github.com/notifications/unsubscribe-auth/AQSTUJJFIRYJAZGVAXRN3232P6ESPAVCNFSM6AAAAABWVDF3RWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRRGA2DKOBTHE>
> 
> pamoroso
>  left a comment 
> (Interlisp/medley#2021)
>  <https://github.com/Interlisp/medley/pull/2021#issuecomment-2661045839>
> I followed that procedure but when loading LLKEY.COM from rmk68 Medley becomes unresponsive.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/2021#issuecomment-2661045839>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJJFIRYJAZGVAXRN3232P6ESPAVCNFSM6AAAAABWVDF3RWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRRGA2DKOBTHE>.
> You are receiving this because you authored the thread.
>

pamoroso commented:
I made the apps loadup from rmk68, switched to rmk65, and opened TEdit from the background menu in a Medley `-f` session.

I see the old arrow key behavior, i.e. only the left and right arrows work but not the up and down ones. And, on the last line of the buffer, I can move right the caret only up to the left of the rightmost character.

pamoroso commented:
I also tried saving a sysout from rmk68 and running the saved sysout from rmk65 but the result is the same: only the left and arrow keys work in TEdit.

rmkaplan commented:
When you switch to rmk68, you need to (LOAD 'TEDIT.LCOM 'SYSLOAD).  Did you do that?

> On Feb 16, 2025, at 4:14 AM, Paolo Amoroso ***@***.***> wrote:
> 
> 
> pamoroso
>  left a comment 
> (Interlisp/medley#2021)
> I made the apps loadup from rmk68, switched to rmk65, and opened TEdit from the background menu in a Medley -f session.
> 
> I see the old arrow key behavior, i.e. only the left and right arrows work but not the up and down ones. And, on the last line of the buffer, I can move right the caret only up to the left of the rightmost character.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.
> 
>  <https://github.com/Interlisp/medley/pull/2021#issuecomment-2661404168> <https://github.com/notifications/unsubscribe-auth/AQSTUJMYFQW5IWDSBT6PTFD2QB6KTAVCNFSM6AAAAABWVDF3RWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRRGQYDIMJWHA>
> 
> pamoroso
>  left a comment 
> (Interlisp/medley#2021)
>  <https://github.com/Interlisp/medley/pull/2021#issuecomment-2661404168>
> I made the apps loadup from rmk68, switched to rmk65, and opened TEdit from the background menu in a Medley -f session.
> 
> I see the old arrow key behavior, i.e. only the left and right arrows work but not the up and down ones. And, on the last line of the buffer, I can move right the caret only up to the left of the rightmost character.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/2021#issuecomment-2661404168>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJMYFQW5IWDSBT6PTFD2QB6KTAVCNFSM6AAAAABWVDF3RWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRRGQYDIMJWHA>.
> You are receiving this because you authored the thread.
>

pamoroso commented:
Do you mean rmk68 when making the sysout with LLKEY?

pamoroso commented:
From rmk65 `(LOAD 'TEDIT.LCOM 'SYSLOAD)` successfully loads TEdit. But then opening TEdit breaks with the error:

```
INTERLISP-ERROR
ARG NOT PARALOOKS
{FMT101/10646:LE-0-0
```

![tedit-error](https://github.com/user-attachments/assets/539c9ed2-f2c8-4c75-9039-9b876b67ae2b)

The context:

```
MOUSE/5(debug)BTV
   MESS1 "ARG NOT PARALOOKS"
   MESS2 {FMT101/10646:LE-0-0}
   NOBREAK NIL
ERROR
\LISPERROR
\DTEST.UFN
   PARALOOKS {FMT101/10646:LE-0-0}
   PC NIL
   TEXTOBJ {TEXTOBJ}#137,45600
   NOSTYLE NIL
   STYLE-SHEET NIL
\TEDIT.APPLY.PARASTYLES
   TEXTOBJ {TEXTOBJ}#137,45600
   PC NIL
   LINE {L95/15784: NIL-NIL}
   IMAGESTREAM #<Output Display Stream/167,114600>
   LINETYPE TRUEDISPLAY
   PLOOKS {FMT101/10646:LE-0-0}
   SCALE NIL
\TEDIT.FORMATLINE.SETUP.PARA
   TEXTOBJ {TEXTOBJ}#137,45600
   CH#1 1
   LINE {L95/15784: NIL-NIL}
   THISLINE {THISLINE}#137,44764
   CHARSLOT NIL
   FONT NIL
   CLOOKS NIL
   TRUEASCENT NIL
   TRUEDESCENT NIL
   LM NIL
   PLOOKS NIL
\TEDIT.FORMATLINE.EMPTY
   TSTREAM #<IO Text Stream/170,117600>
   CH#1 1
   LINE {L95/15784: NIL-NIL}
   REGION NIL
   IMAGESTREAM #<Output Display Stream/167,114600>
   FORMATTINGSTATE NIL
   TEXTOBJ {TEXTOBJ}#137,45600
   OFFSET 0
   TRUEASCENT -1
   TRUEDESCENT -1
   ASCENTB 0
   DESCENTB 0
   ASCENTC 0
   DESCENTC 0
   OVERHANG 0
   SPACELEFT 0
   TX 0
   BOXSTREAM #<Output Display Stream/167,114600>
   THISLINE {THISLINE}#137,44764
   LINETYPE NIL
   WIDTH NIL
   WMARGIN NIL
   SCALE NIL
   FMTSPEC NIL
   RIGHTMARGIN NIL
   HASKERN NIL
   PC NIL
   CHARSLOT NIL
   PREVSP NIL
   1STLN NIL
   CHNOB NIL
   FORCED-END NIL
   CHNO NIL
   LX1 NIL
   TX NIL
   TXB NIL
   FONT NIL
   CHARSLOTB NIL
   TABPENDING NIL
   PREVHYPH NIL
   PREVDHYPH NIL
   START-OF-PIECE 1
   UNBREAKABLE NIL
   OLDPIECE {PIECE}#137,42744
   OLDPCCHARSLEFT 0
   OLDCARETLOOKS {CL100/19920:Gacha10}
\TEDIT.FORMATLINE
   PANE {WINDOW}#157,26074
   TEXTOBJ {TEXTOBJ}#137,45600
   PREVLINE {L95/15826: 0-0 FED}
\TEDIT.SUFFIXLINE.CREATE
   TEXTOBJ {TEXTOBJ}#137,45600
   PANE {WINDOW}#157,26074
   LCHARLAST NIL
   YBOT NIL
   PREFIX {L95/15826: 0-0 FED}
\TEDIT.PANE.CREATELINES
   PANE {WINDOW}#157,26074
   TSTREAM #<IO Text Stream/170,117600>
   PROPS (BEING-EDITED T)
   AFTERPANE NIL
   LCHAR1 NIL
   TEXTOBJ {TEXTOBJ}#137,45600
   MENUPROP NIL
   SEL {SEL: unset 100/33518}
   LASTVISIBLE NIL
\TEDIT.WINDOW.SETUP
   WINDOW {WINDOW}#157,26074
   TSTREAM #<IO Text Stream/170,117600>
   PROPS (BEING-EDITED T)
   TEXTOBJ {TEXTOBJ}#137,45600
   FILEPTR 0
\TEDIT.OPENTEXTSTREAM.WINDOW
   SI::*CLEANUP-FORMS* SI::RESETUNWIND
   TSTREAM #<IO Text Stream/170,117600>
   TEXTOBJ {TEXTOBJ}#137,45600
   TEDIT.GET.FINISHEDFORMS NIL
   PRIMPANE NIL
   START NIL
SI::*UNWIND-PROTECT*
   TEXT NIL
   WINDOW {WINDOW}#157,26074
   START/PROPS NIL
   END NIL
   PROPS (BEING-EDITED T)
   LISPXHIST NIL
   SI::*RESETFORMS* NIL
   RESETSTATE NIL
OPENTEXTSTREAM
   TEXT NIL
   WINDOW NIL
   DONTSPAWN NIL
   PROPS (BEING-EDITED T)
   TSTREAM NIL
   PROC NIL
TEDIT
   *FORM* (TEDIT)
   *ARGVAL* NIL
   *TAIL* NIL
   *FN* TEDIT
\EVALFORM
   \INTERNAL NIL
EVAL
   SI::*DUMMY-FOR-CATCH* T
   SI::*CATCH-RETURN-FROM* (&)
DOBACKGROUNDCOMA0010
   FORM (TEDIT)
   SI::NLSETQ-VALUE NIL
   *PROCEED-CASES* (&)
   SI::*NLSETQFLAG* NIL
DOBACKGROUNDCOM
   ARGS 1
   WINDOW NIL
DOWINDOWCOM
   \MHCOM NIL
   \MHPROCESS NIL
   \MHWINDOW NIL
   \MOUSEBUSY T
WINDOW.MOUSE.HANDLER
   \OLDTTY #<Process EXEC/174,25204>
   \MOUSEBUSY NIL
\MOUSE.PROCESS
   *FORM* (\MOUSE.PROCESS)
   *ARGVAL* NIL
   *TAIL* NIL
   *FN* \MOUSE.PROCESS
\EVALFORM
   %#FORM# (\MOUSE.PROCESS)
   *CURRENT-PROCESS* #<Process OLDMOUSE/174,25410>
   HELPFLAG BREAK!
   \CURRENTDISPLAYLINE 0
   \#DISPLAYLINES 40
   \LINEBUF.OFD #<IO Stream on T/174,74500>
   *READTABLE* #<ReadTable INTERLISP/174,73714>
   \PRIMTERMTABLE {TERMTABLEP}#174,66740
   \PRIMTERMSA {CHARTABLE}#174,67000
   TtyDisplayStream 
#<Output Display Stream/146,22000>
   SI::*RESETFORMS* NIL
   \INTERRUPTABLE T
   \TTYWINDOW NIL
   READBUF NIL
   \TERM.OFD #<Output Stream on T/170,117200>
   *STANDARD-OUTPUT* #<Output Stream on T/170,117200>
   *STANDARD-INPUT* #<IO Stream on T/174,74500>
\MAKE.PROCESS0
T
```

rmkaplan commented:
Try first (LOAD 'TEDIT-EXPORTS.ALL) before the TEDIT-DEBUG.    If that doesn't work, then I think that 68 has to be merged first.  Maybe you can merge 68 into 65?

I thought 68 was a small independent change that should therefore be pushed separately, on the theory that that's the right way of dealing with git.  But maybe not.

> On Feb 16, 2025, at 8:51 AM, Paolo Amoroso ***@***.***> wrote:
> 
> 
> pamoroso
>  left a comment 
> (Interlisp/medley#2021)
> From rmk65 (LOAD 'TEDIT.LCOM 'SYSLOAD) successfully loads TEdit. But then opening TEdit breaks with the error:
> 
> INTERLISP-ERROR
> ARG NOT PARALOOKS
> {FMT101/10646:LE-0-0
> tedit-error.png (view on web)
> 
> The context:
> 
> MOUSE/5(debug)BTV
>    MESS1 "ARG NOT PARALOOKS"
>    MESS2 {FMT101/10646:LE-0-0}
>    NOBREAK NIL
> ERROR
> \LISPERROR
> \DTEST.UFN
>    PARALOOKS {FMT101/10646:LE-0-0}
>    PC NIL
>    TEXTOBJ {TEXTOBJ}#137,45600
>    NOSTYLE NIL
>    STYLE-SHEET NIL
> \TEDIT.APPLY.PARASTYLES
>    TEXTOBJ {TEXTOBJ}#137,45600
>    PC NIL
>    LINE {L95/15784: NIL-NIL}
>    IMAGESTREAM #<Output Display Stream/167,114600>
>    LINETYPE TRUEDISPLAY
>    PLOOKS {FMT101/10646:LE-0-0}
>    SCALE NIL
> \TEDIT.FORMATLINE.SETUP.PARA
>    TEXTOBJ {TEXTOBJ}#137,45600
>    CH#1 1
>    LINE {L95/15784: NIL-NIL}
>    THISLINE {THISLINE}#137,44764
>    CHARSLOT NIL
>    FONT NIL
>    CLOOKS NIL
>    TRUEASCENT NIL
>    TRUEDESCENT NIL
>    LM NIL
>    PLOOKS NIL
> \TEDIT.FORMATLINE.EMPTY
>    TSTREAM #<IO Text Stream/170,117600>
>    CH#1 1
>    LINE {L95/15784: NIL-NIL}
>    REGION NIL
>    IMAGESTREAM #<Output Display Stream/167,114600>
>    FORMATTINGSTATE NIL
>    TEXTOBJ {TEXTOBJ}#137,45600
>    OFFSET 0
>    TRUEASCENT -1
>    TRUEDESCENT -1
>    ASCENTB 0
>    DESCENTB 0
>    ASCENTC 0
>    DESCENTC 0
>    OVERHANG 0
>    SPACELEFT 0
>    TX 0
>    BOXSTREAM #<Output Display Stream/167,114600>
>    THISLINE {THISLINE}#137,44764
>    LINETYPE NIL
>    WIDTH NIL
>    WMARGIN NIL
>    SCALE NIL
>    FMTSPEC NIL
>    RIGHTMARGIN NIL
>    HASKERN NIL
>    PC NIL
>    CHARSLOT NIL
>    PREVSP NIL
>    1STLN NIL
>    CHNOB NIL
>    FORCED-END NIL
>    CHNO NIL
>    LX1 NIL
>    TX NIL
>    TXB NIL
>    FONT NIL
>    CHARSLOTB NIL
>    TABPENDING NIL
>    PREVHYPH NIL
>    PREVDHYPH NIL
>    START-OF-PIECE 1
>    UNBREAKABLE NIL
>    OLDPIECE {PIECE}#137,42744
>    OLDPCCHARSLEFT 0
>    OLDCARETLOOKS {CL100/19920:Gacha10}
> \TEDIT.FORMATLINE
>    PANE {WINDOW}#157,26074
>    TEXTOBJ {TEXTOBJ}#137,45600
>    PREVLINE {L95/15826: 0-0 FED}
> \TEDIT.SUFFIXLINE.CREATE
>    TEXTOBJ {TEXTOBJ}#137,45600
>    PANE {WINDOW}#157,26074
>    LCHARLAST NIL
>    YBOT NIL
>    PREFIX {L95/15826: 0-0 FED}
> \TEDIT.PANE.CREATELINES
>    PANE {WINDOW}#157,26074
>    TSTREAM #<IO Text Stream/170,117600>
>    PROPS (BEING-EDITED T)
>    AFTERPANE NIL
>    LCHAR1 NIL
>    TEXTOBJ {TEXTOBJ}#137,45600
>    MENUPROP NIL
>    SEL {SEL: unset 100/33518}
>    LASTVISIBLE NIL
> \TEDIT.WINDOW.SETUP
>    WINDOW {WINDOW}#157,26074
>    TSTREAM #<IO Text Stream/170,117600>
>    PROPS (BEING-EDITED T)
>    TEXTOBJ {TEXTOBJ}#137,45600
>    FILEPTR 0
> \TEDIT.OPENTEXTSTREAM.WINDOW
>    SI::*CLEANUP-FORMS* SI::RESETUNWIND
>    TSTREAM #<IO Text Stream/170,117600>
>    TEXTOBJ {TEXTOBJ}#137,45600
>    TEDIT.GET.FINISHEDFORMS NIL
>    PRIMPANE NIL
>    START NIL
> SI::*UNWIND-PROTECT*
>    TEXT NIL
>    WINDOW {WINDOW}#157,26074
>    START/PROPS NIL
>    END NIL
>    PROPS (BEING-EDITED T)
>    LISPXHIST NIL
>    SI::*RESETFORMS* NIL
>    RESETSTATE NIL
> OPENTEXTSTREAM
>    TEXT NIL
>    WINDOW NIL
>    DONTSPAWN NIL
>    PROPS (BEING-EDITED T)
>    TSTREAM NIL
>    PROC NIL
> TEDIT
>    *FORM* (TEDIT)
>    *ARGVAL* NIL
>    *TAIL* NIL
>    *FN* TEDIT
> \EVALFORM
>    \INTERNAL NIL
> EVAL
>    SI::*DUMMY-FOR-CATCH* T
>    SI::*CATCH-RETURN-FROM* (&)
> DOBACKGROUNDCOMA0010
>    FORM (TEDIT)
>    SI::NLSETQ-VALUE NIL
>    *PROCEED-CASES* (&)
>    SI::*NLSETQFLAG* NIL
> DOBACKGROUNDCOM
>    ARGS 1
>    WINDOW NIL
> DOWINDOWCOM
>    \MHCOM NIL
>    \MHPROCESS NIL
>    \MHWINDOW NIL
>    \MOUSEBUSY T
> WINDOW.MOUSE.HANDLER
>    \OLDTTY #<Process EXEC/174,25204>
>    \MOUSEBUSY NIL
> \MOUSE.PROCESS
>    *FORM* (\MOUSE.PROCESS)
>    *ARGVAL* NIL
>    *TAIL* NIL
>    *FN* \MOUSE.PROCESS
> \EVALFORM
>    %#FORM# (\MOUSE.PROCESS)
>    *CURRENT-PROCESS* #<Process OLDMOUSE/174,25410>
>    HELPFLAG BREAK!
>    \CURRENTDISPLAYLINE 0
>    \#DISPLAYLINES 40
>    \LINEBUF.OFD #<IO Stream on T/174,74500>
>    *READTABLE* #<ReadTable INTERLISP/174,73714>
>    \PRIMTERMTABLE {TERMTABLEP}#174,66740
>    \PRIMTERMSA {CHARTABLE}#174,67000
>    TtyDisplayStream 
> #<Output Display Stream/146,22000>
>    SI::*RESETFORMS* NIL
>    \INTERRUPTABLE T
>    \TTYWINDOW NIL
>    READBUF NIL
>    \TERM.OFD #<Output Stream on T/170,117200>
>    *STANDARD-OUTPUT* #<Output Stream on T/170,117200>
>    *STANDARD-INPUT* #<IO Stream on T/174,74500>
> \MAKE.PROCESS0
> T
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.
> 
>  <https://github.com/user-attachments/assets/539c9ed2-f2c8-4c75-9039-9b876b67ae2b> <https://github.com/Interlisp/medley/pull/2021#issuecomment-2661521246> <https://github.com/notifications/unsubscribe-auth/AQSTUJNV6YQ7A4MI4WDHEG32QC6ZHAVCNFSM6AAAAABWVDF3RWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRRGUZDCMRUGY>
> 
> pamoroso
>  left a comment 
> (Interlisp/medley#2021)
>  <https://github.com/Interlisp/medley/pull/2021#issuecomment-2661521246>
> From rmk65 (LOAD 'TEDIT.LCOM 'SYSLOAD) successfully loads TEdit. But then opening TEdit breaks with the error:
> 
> INTERLISP-ERROR
> ARG NOT PARALOOKS
> {FMT101/10646:LE-0-0
> tedit-error.png (view on web) <https://github.com/user-attachments/assets/539c9ed2-f2c8-4c75-9039-9b876b67ae2b>
> The context:
> 
> MOUSE/5(debug)BTV
>    MESS1 "ARG NOT PARALOOKS"
>    MESS2 {FMT101/10646:LE-0-0}
>    NOBREAK NIL
> ERROR
> \LISPERROR
> \DTEST.UFN
>    PARALOOKS {FMT101/10646:LE-0-0}
>    PC NIL
>    TEXTOBJ {TEXTOBJ}#137,45600
>    NOSTYLE NIL
>    STYLE-SHEET NIL
> \TEDIT.APPLY.PARASTYLES
>    TEXTOBJ {TEXTOBJ}#137,45600
>    PC NIL
>    LINE {L95/15784: NIL-NIL}
>    IMAGESTREAM #<Output Display Stream/167,114600>
>    LINETYPE TRUEDISPLAY
>    PLOOKS {FMT101/10646:LE-0-0}
>    SCALE NIL
> \TEDIT.FORMATLINE.SETUP.PARA
>    TEXTOBJ {TEXTOBJ}#137,45600
>    CH#1 1
>    LINE {L95/15784: NIL-NIL}
>    THISLINE {THISLINE}#137,44764
>    CHARSLOT NIL
>    FONT NIL
>    CLOOKS NIL
>    TRUEASCENT NIL
>    TRUEDESCENT NIL
>    LM NIL
>    PLOOKS NIL
> \TEDIT.FORMATLINE.EMPTY
>    TSTREAM #<IO Text Stream/170,117600>
>    CH#1 1
>    LINE {L95/15784: NIL-NIL}
>    REGION NIL
>    IMAGESTREAM #<Output Display Stream/167,114600>
>    FORMATTINGSTATE NIL
>    TEXTOBJ {TEXTOBJ}#137,45600
>    OFFSET 0
>    TRUEASCENT -1
>    TRUEDESCENT -1
>    ASCENTB 0
>    DESCENTB 0
>    ASCENTC 0
>    DESCENTC 0
>    OVERHANG 0
>    SPACELEFT 0
>    TX 0
>    BOXSTREAM #<Output Display Stream/167,114600>
>    THISLINE {THISLINE}#137,44764
>    LINETYPE NIL
>    WIDTH NIL
>    WMARGIN NIL
>    SCALE NIL
>    FMTSPEC NIL
>    RIGHTMARGIN NIL
>    HASKERN NIL
>    PC NIL
>    CHARSLOT NIL
>    PREVSP NIL
>    1STLN NIL
>    CHNOB NIL
>    FORCED-END NIL
>    CHNO NIL
>    LX1 NIL
>    TX NIL
>    TXB NIL
>    FONT NIL
>    CHARSLOTB NIL
>    TABPENDING NIL
>    PREVHYPH NIL
>    PREVDHYPH NIL
>    START-OF-PIECE 1
>    UNBREAKABLE NIL
>    OLDPIECE {PIECE}#137,42744
>    OLDPCCHARSLEFT 0
>    OLDCARETLOOKS {CL100/19920:Gacha10}
> \TEDIT.FORMATLINE
>    PANE {WINDOW}#157,26074
>    TEXTOBJ {TEXTOBJ}#137,45600
>    PREVLINE {L95/15826: 0-0 FED}
> \TEDIT.SUFFIXLINE.CREATE
>    TEXTOBJ {TEXTOBJ}#137,45600
>    PANE {WINDOW}#157,26074
>    LCHARLAST NIL
>    YBOT NIL
>    PREFIX {L95/15826: 0-0 FED}
> \TEDIT.PANE.CREATELINES
>    PANE {WINDOW}#157,26074
>    TSTREAM #<IO Text Stream/170,117600>
>    PROPS (BEING-EDITED T)
>    AFTERPANE NIL
>    LCHAR1 NIL
>    TEXTOBJ {TEXTOBJ}#137,45600
>    MENUPROP NIL
>    SEL {SEL: unset 100/33518}
>    LASTVISIBLE NIL
> \TEDIT.WINDOW.SETUP
>    WINDOW {WINDOW}#157,26074
>    TSTREAM #<IO Text Stream/170,117600>
>    PROPS (BEING-EDITED T)
>    TEXTOBJ {TEXTOBJ}#137,45600
>    FILEPTR 0
> \TEDIT.OPENTEXTSTREAM.WINDOW
>    SI::*CLEANUP-FORMS* SI::RESETUNWIND
>    TSTREAM #<IO Text Stream/170,117600>
>    TEXTOBJ {TEXTOBJ}#137,45600
>    TEDIT.GET.FINISHEDFORMS NIL
>    PRIMPANE NIL
>    START NIL
> SI::*UNWIND-PROTECT*
>    TEXT NIL
>    WINDOW {WINDOW}#157,26074
>    START/PROPS NIL
>    END NIL
>    PROPS (BEING-EDITED T)
>    LISPXHIST NIL
>    SI::*RESETFORMS* NIL
>    RESETSTATE NIL
> OPENTEXTSTREAM
>    TEXT NIL
>    WINDOW NIL
>    DONTSPAWN NIL
>    PROPS (BEING-EDITED T)
>    TSTREAM NIL
>    PROC NIL
> TEDIT
>    *FORM* (TEDIT)
>    *ARGVAL* NIL
>    *TAIL* NIL
>    *FN* TEDIT
> \EVALFORM
>    \INTERNAL NIL
> EVAL
>    SI::*DUMMY-FOR-CATCH* T
>    SI::*CATCH-RETURN-FROM* (&)
> DOBACKGROUNDCOMA0010
>    FORM (TEDIT)
>    SI::NLSETQ-VALUE NIL
>    *PROCEED-CASES* (&)
>    SI::*NLSETQFLAG* NIL
> DOBACKGROUNDCOM
>    ARGS 1
>    WINDOW NIL
> DOWINDOWCOM
>    \MHCOM NIL
>    \MHPROCESS NIL
>    \MHWINDOW NIL
>    \MOUSEBUSY T
> WINDOW.MOUSE.HANDLER
>    \OLDTTY #<Process EXEC/174,25204>
>    \MOUSEBUSY NIL
> \MOUSE.PROCESS
>    *FORM* (\MOUSE.PROCESS)
>    *ARGVAL* NIL
>    *TAIL* NIL
>    *FN* \MOUSE.PROCESS
> \EVALFORM
>    %#FORM# (\MOUSE.PROCESS)
>    *CURRENT-PROCESS* #<Process OLDMOUSE/174,25410>
>    HELPFLAG BREAK!
>    \CURRENTDISPLAYLINE 0
>    \#DISPLAYLINES 40
>    \LINEBUF.OFD #<IO Stream on T/174,74500>
>    *READTABLE* #<ReadTable INTERLISP/174,73714>
>    \PRIMTERMTABLE {TERMTABLEP}#174,66740
>    \PRIMTERMSA {CHARTABLE}#174,67000
>    TtyDisplayStream 
> #<Output Display Stream/146,22000>
>    SI::*RESETFORMS* NIL
>    \INTERRUPTABLE T
>    \TTYWINDOW NIL
>    READBUF NIL
>    \TERM.OFD #<Output Stream on T/170,117200>
>    *STANDARD-OUTPUT* #<Output Stream on T/170,117200>
>    *STANDARD-INPUT* #<IO Stream on T/174,74500>
> \MAKE.PROCESS0
> T
> —
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/2021#issuecomment-2661521246>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJNV6YQ7A4MI4WDHEG32QC6ZHAVCNFSM6AAAAABWVDF3RWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRRGUZDCMRUGY>.
> You are receiving this because you authored the thread.
>

pamoroso commented:
I did `(LOAD 'TEDIT-EXPORTS.ALL)` and then `(LOAD 'TEDIT.LCOM 'SYSLOAD)` but I still get the `ARG NOT PARALOOKS` error when opening TEdit.

I have no idea how to merge the branches, also considering that I'm working on a direct clone of the Medley repo (I didn't expect to change it myself) rather than a clone of a fork.

rmkaplan commented:
I'm in the dark too, but I think I will just include the LLKEY changes in 65.  That's where we eventually want to end up, I can back out on the Lisp side, without depending on the git configuration, if people object to the simple LLKEY change.

What I was trying to avoid was having merge conflicts at the end, which I wouldn't know how to unwind.  

I pushed a new 65, if that works and noone objects to LLKEY, then we should abandon rmk68.

> On Feb 16, 2025, at 9:08 AM, Paolo Amoroso ***@***.***> wrote:
> 
> 
> pamoroso
>  left a comment 
> (Interlisp/medley#2021)
> I did (LOAD 'TEDIT-EXPORTS.ALL) and then (LOAD 'TEDIT.LCOM 'SYSLOAD) but I still get the ARG NOT PARALOOKS error when opening TEdit.
> 
> I have no idea how to merge the branches, also considering that I'm working on a direct clone of the Medley repo (I didn't expect to change it myself) rather than a clone of a fork.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.
> 
>  <https://github.com/Interlisp/medley/pull/2021#issuecomment-2661528535> <https://github.com/notifications/unsubscribe-auth/AQSTUJI6JYW5JLDL7QVX7YD2QDAZJAVCNFSM6AAAAABWVDF3RWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRRGUZDQNJTGU>
> 
> pamoroso
>  left a comment 
> (Interlisp/medley#2021)
>  <https://github.com/Interlisp/medley/pull/2021#issuecomment-2661528535>
> I did (LOAD 'TEDIT-EXPORTS.ALL) and then (LOAD 'TEDIT.LCOM 'SYSLOAD) but I still get the ARG NOT PARALOOKS error when opening TEdit.
> 
> I have no idea how to merge the branches, also considering that I'm working on a direct clone of the Medley repo (I didn't expect to change it myself) rather than a clone of a fork.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/2021#issuecomment-2661528535>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJI6JYW5JLDL7QVX7YD2QDAZJAVCNFSM6AAAAABWVDF3RWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRRGUZDQNJTGU>.
> You are receiving this because you authored the thread.
>

nbriggs commented on sources/LLKEY:
```diff
@@ -1754,13 +1752,13 @@
         (75 ("^B" "^B" NOLOCKSHIFT))
         (81 ("Function,FF" "7" NOLOCKSHIFT))
-        (82 ("#-" "8" NOLOCKSHIFT))
+        (UPARROW ("Meta,^" "8" NOLOCKSHIFT))
```
No other keyactions are specified by key name - should these not be the key number?

pamoroso commented:
With 6f34965 I can finally test the new arrow key behavior in TEdit.

All cursor keys work and I can now move the caret immediately to the right of the rightmost character of the last line of a buffer.

Moving up and down between lines moves the caret to the expected spot with monospace fonts like Gacha or with proportional fonts like Helvetica when the lines have the same font. When the lines widely differ in font and size (e.g. Gacha 8 and Times Roman 36) the caret moves to unexpected spots (understandably) but still usefully close to where I'd like.

pamoroso commented:
One particularly useful thing I hadn't noticed so far. In the text input fields of free menus, such as the `Size:` field of TEdit's character looks menu, I can move the caret left and right with the corresponding arrow keys. That's handy for quick edits.

rmkaplan commented:
These are all interpreted by a call to KEYACTION, which knows how to decode keynames.  I think it would have been (and would be) better if all of these tables had been specified in terms of keynames, since then there would be only one place where the mapping from name to bit-number is defined, the variable \KEYNAMES>

> On Feb 16, 2025, at 9:26 AM, Nick Briggs ***@***.***> wrote:
> 
> 
> @nbriggs commented on this pull request.
> 
> In sources/LLKEY <https://github.com/Interlisp/medley/pull/2021#discussion_r1957374861>:
> 
> > @@ -1754,13 +1752,13 @@
>          (74 ("Function,LF" "Function,LF" NOLOCKSHIFT))
>          (75 ("^B" "^B" NOLOCKSHIFT))
>          (81 ("Function,FF" "7" NOLOCKSHIFT))
> -        (82 ("#-" "8" NOLOCKSHIFT))
> +        (UPARROW ("Meta,^" "8" NOLOCKSHIFT))
> No other keyactions are specified by key name - should these not be the key number?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/2021#pullrequestreview-2619679610>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJPKZ2YG5NHBL5QF5XD2QDC5TAVCNFSM6AAAAABWVDF3RWVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMMJZGY3TSNRRGA>.
> You are receiving this because you authored the thread.
>

nbriggs commented:
Except `\KEYNAMES` only implicitly references a bit number by the position within the list - which is certainly a source of possible error and does mean that you have to process it to produce anything useful for maiko or anything lower level where you need the bit number.

nbriggs commented:
IL:NTH (really NTH tail) indexes from 1, while CL:NTH indexes from 0 - and who would know whether the bits in the keyboard bit vector are origin 0 or 1 when looking at `\KEYNAMES` (the Maiko code and the lower level Lisp code that does the bit-bashing use 0-origin numbering).

rmkaplan commented:
I think \KEYNAMETONUMBER and \KEYNUMBERTONAME are the official ways to do the mappings.

> On Feb 16, 2025, at 11:00 AM, Nick Briggs ***@***.***> wrote:
> 
> 
> nbriggs
>  left a comment 
> (Interlisp/medley#2021)
> IL:NTH (really NTH tail) indexes from 1, while CL:NTH indexes from 0 - and who would know whether the bits in the keyboard bit vector are origin 0 or 1 when looking at \KEYNAMES (the Maiko code and the lower level Lisp code that does the bit-bashing use 0-origin numbering).
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.
> 
>  <https://github.com/Interlisp/medley/pull/2021#issuecomment-2661570739> <https://github.com/notifications/unsubscribe-auth/AQSTUJILWAZA46SSZB7XYDD2QDN4HAVCNFSM6AAAAABWVDF3RWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRRGU3TANZTHE>
> 
> nbriggs
>  left a comment 
> (Interlisp/medley#2021)
>  <https://github.com/Interlisp/medley/pull/2021#issuecomment-2661570739>
> IL:NTH (really NTH tail) indexes from 1, while CL:NTH indexes from 0 - and who would know whether the bits in the keyboard bit vector are origin 0 or 1 when looking at \KEYNAMES (the Maiko code and the lower level Lisp code that does the bit-bashing use 0-origin numbering).
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/2021#issuecomment-2661570739>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJILWAZA46SSZB7XYDD2QDN4HAVCNFSM6AAAAABWVDF3RWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRRGU3TANZTHE>.
> You are receiving this because you authored the thread.
>

rmkaplan commented:
What is the unexpected behavior, can you give an example?

It tries to put the caret at the position closest to the original X position.  So, moving from a small font to a line with a large font, it should go to the left or right of the big character depending on which end is closest to the original small-character position.

> On Feb 16, 2025, at 9:40 AM, Paolo Amoroso ***@***.***> wrote:
> 
> When the lines widely differ in font and size (e.g. Gacha 8 and Times Roman 36) the caret moves to unexpected spots (understandably) but still usefully close to where I'd like.

pamoroso commented:
Sometimes moving from a small font to a line with a large font goes a bit more to the left of the big character than it would seem given what's closest to the original small-character position.

But, in general, the spot the caret moves to is good enough. It's difficult to form clear expectations based on the typefaces and sizes.

nbriggs commented:
Moving to the gap in the next/previous line that's nearest in X position is what, for example, TextEdit on the mac does - set up a 10pt line of text below a 36pt line of text and use up/down to move around to see an example.

rmkaplan commented:
Does it end up between the characters that you would expect it to end up between?

The actual blinking position should be the same as if you click between those characters.

> On Feb 16, 2025, at 12:14 PM, Paolo Amoroso ***@***.***> wrote:
> 
> 
> pamoroso
>  left a comment 
> (Interlisp/medley#2021)
> Sometimes moving from a small font to a line with a large font goes a bit more to the left of the big character than it would seem given what's closest to the original small-character position.
> 
> But, in general, the spot the caret moves to is good enough. It's difficult to form clear expectations based on the typefaces and sizes.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.
> 
>  <https://github.com/Interlisp/medley/pull/2021#issuecomment-2661597544> <https://github.com/notifications/unsubscribe-auth/AQSTUJLGR6CUHC4LF2CR6RD2QDWS3AVCNFSM6AAAAABWVDF3RWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRRGU4TONJUGQ>
> 
> pamoroso
>  left a comment 
> (Interlisp/medley#2021)
>  <https://github.com/Interlisp/medley/pull/2021#issuecomment-2661597544>
> Sometimes moving from a small font to a line with a large font goes a bit more to the left of the big character than it would seem given what's closest to the original small-character position.
> 
> But, in general, the spot the caret moves to is good enough. It's difficult to form clear expectations based on the typefaces and sizes.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/2021#issuecomment-2661597544>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJLGR6CUHC4LF2CR6RD2QDWS3AVCNFSM6AAAAABWVDF3RWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRRGU4TONJUGQ>.
> You are receiving this because you authored the thread.
>

rmkaplan commented:
I put in some checks to stop up/down/left/right arrows from moving outside of the menu type-in fields.  You can go back and forth in a given field, but not jump to another field on the same line or a different line.

pamoroso commented:
I tested the changes up to 0485652 and everything seems to be working fine. Moving left and right with the arrow keys in a TEdit text field stays within the field.

As for moving up and down between lines, suppose in TEdit you have the following lines of default Gacha 10:

```
aaaa aaaa aaaa
aaaa aaaa aaaa
```

Start with the caret here:

```
aaaa aaaa aaaa
aaaa aaaa ^aaaa
```

Move up:

```
aaaa aaaa ^aaaa
aaaa aaaa aaaa
```

Move to the beginning of the string to the left:

```
aaaa ^aaaa aaaa
aaaa aaaa aaaa
```

If you now move down the caret ends up here:

```
aaaa aaaa aaaa
aaaa aaaa ^aaaa
```

whereas I would have expected here:

```
aaaa aaaa aaaa
aaaa ^aaaa aaaa
```

pamoroso commented:
The screen recording in this .zip archive shows an example of a mismatch between my expectation on where the caret should move when pressing an up or down arrow key and where it actually ends up in TEdit:  [tedit-cursor-keys.webm.zip](https://github.com/user-attachments/files/18825811/tedit-cursor-keys.webm.zip)

The heading is Times Roman 36 bold. The caret starts here:

```
Documen^t Title
```

and ends up here when pressing the down arrow:

```
... first line of^ the first...
```

I expected the caret to land somewhere closer to here instead:

```
... first line^ of the first...
```

rmkaplan commented:
I think this is an interaction betwen the move up/down and move left/right.   The request for up/down at the meeting last week was for it to remember the caret position before the first up/down in a sequence of up/downs.  So that if the next target line is short but the one after that is long, it will revert to the original position.

I set it so that it remembers that position until the next mouse click.  I think this example indicates that left/right should also cause it to forgeet the original position.  I pushed a fix, try it now.



> On Feb 17, 2025, at 2:56 AM, Paolo Amoroso ***@***.***> wrote:
> 
> 
> pamoroso
>  left a comment 
> (Interlisp/medley#2021)
> I tested the changes up to 0485652 and everything seems to be working fine. Moving left and right with the arrow keys in a TEdit text field stays within the field.
> 
> As for moving up and down between lines, suppose in TEdit you have the following lines of default Gacha 10:
> 
> aaaa aaaa aaaa
> aaaa aaaa aaaa
> Start with the caret here:
> 
> aaaa aaaa aaaa
> aaaa aaaa ^aaaa
> Move up:
> 
> aaaa aaaa ^aaaa
> aaaa aaaa aaaa
> Move to the beginning of the string to the left:
> 
> aaaa ^aaaa aaaa
> aaaa aaaa aaaa
> If you now move down the caret ends up here:
> 
> aaaa aaaa aaaa
> aaaa aaaa ^aaaa
> whereas I would have expected here:
> 
> aaaa aaaa aaaa
> aaaa ^aaaa aaaa
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.
> 
>  <https://github.com/Interlisp/medley/commit/048565263760009321bf0e3c10906e62265269ce> <https://github.com/Interlisp/medley/pull/2021#issuecomment-2662752250> <https://github.com/notifications/unsubscribe-auth/AQSTUJKX4IWH4YO5LPNNEKT2QG54VAVCNFSM6AAAAABWVDF3RWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRSG42TEMRVGA>
> 
> pamoroso
>  left a comment 
> (Interlisp/medley#2021)
>  <https://github.com/Interlisp/medley/pull/2021#issuecomment-2662752250>
> I tested the changes up to 0485652 <https://github.com/Interlisp/medley/commit/048565263760009321bf0e3c10906e62265269ce> and everything seems to be working fine. Moving left and right with the arrow keys in a TEdit text field stays within the field.
> 
> As for moving up and down between lines, suppose in TEdit you have the following lines of default Gacha 10:
> 
> aaaa aaaa aaaa
> aaaa aaaa aaaa
> Start with the caret here:
> 
> aaaa aaaa aaaa
> aaaa aaaa ^aaaa
> Move up:
> 
> aaaa aaaa ^aaaa
> aaaa aaaa aaaa
> Move to the beginning of the string to the left:
> 
> aaaa ^aaaa aaaa
> aaaa aaaa aaaa
> If you now move down the caret ends up here:
> 
> aaaa aaaa aaaa
> aaaa aaaa ^aaaa
> whereas I would have expected here:
> 
> aaaa aaaa aaaa
> aaaa ^aaaa aaaa
> —
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/2021#issuecomment-2662752250>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJKX4IWH4YO5LPNNEKT2QG54VAVCNFSM6AAAAABWVDF3RWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRSG42TEMRVGA>.
> You are receiving this because you authored the thread.
>

rmkaplan commented:
Hmm, I don't see this.  I'll poke around some more.

> On Feb 17, 2025, at 5:22 AM, Paolo Amoroso ***@***.***> wrote:
> 
> 
> pamoroso
>  left a comment 
> (Interlisp/medley#2021)
> The screen recording in this .zip archive shows an example of a mismatch between my expectation on where the caret should move when pressing an up or down arrow key and where it actually ends up in TEdit: tedit-cursor-keys.webm.zip
> 
> The heading is Times Roman 36 bold. The caret starts here:
> 
> Documen^t Title
> and ends up here when pressing the down arrow:
> 
> ... first line of^ the first...
> I expected the caret to land somewhere closer to here instead:
> 
> ... first line^ of the first...
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.
> 
>  <https://github.com/user-attachments/files/18825811/tedit-cursor-keys.webm.zip> <https://github.com/Interlisp/medley/pull/2021#issuecomment-2663126375> <https://github.com/notifications/unsubscribe-auth/AQSTUJMPMXPCC4D6HUXU6TL2QHPCBAVCNFSM6AAAAABWVDF3RWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRTGEZDMMZXGU>
> 
> pamoroso
>  left a comment 
> (Interlisp/medley#2021)
>  <https://github.com/Interlisp/medley/pull/2021#issuecomment-2663126375>
> The screen recording in this .zip archive shows an example of a mismatch between my expectation on where the caret should move when pressing an up or down arrow key and where it actually ends up in TEdit: tedit-cursor-keys.webm.zip <https://github.com/user-attachments/files/18825811/tedit-cursor-keys.webm.zip>
> The heading is Times Roman 36 bold. The caret starts here:
> 
> Documen^t Title
> and ends up here when pressing the down arrow:
> 
> ... first line of^ the first...
> I expected the caret to land somewhere closer to here instead:
> 
> ... first line^ of the first...
> —
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/2021#issuecomment-2663126375>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJMPMXPCC4D6HUXU6TL2QHPCBAVCNFSM6AAAAABWVDF3RWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRTGEZDMMZXGU>.
> You are receiving this because you authored the thread.
>

pamoroso commented:
Pressing the down arrow key when TEdit's `Get` and `Put` commands request a file name in the prompt area inserts `^J` in the area. Pressing the other arrow keys flashes the screen as I'd expect for the down arrow too.

pamoroso commented:
To clarify, the left and right arrows correctly flash the screen only when the caret hits the boundaries of the prompt area or the area is empty, otherwise the keys make the caret move orizontally as expected. The arrow key behavior is the same in the prompt areas of other programs such as the File Browser and the CLOS browser.

pamoroso commented:
I tested the changes up to 47bf5a0 and I don't have anything unusual to report.

pamoroso commented:
On this branch pressing the arrow keys in the input field of `PROMPTFORWORD` inserts the following:

* left: `#1,60`
* right: `#1,62`
* up: `#1,94`
* down: `#1,10`

Here I pressed left, right, up, down:

![promptforword](https://github.com/user-attachments/assets/18c221a0-9cb2-47ba-8834-152eae3b4e3c)

On Medley Online nothing happens when pressing the same keys in the `PROMPTFORWORD` field.

nbriggs reviewed (approved):
Approved in meeting 24 Feb 2025.  Any additional issues that show up should get new issues.

pamoroso commented:
I built Medley master that includes the merged #2021 and now the arrow keys insert black boxes in both the Exec and TEdit.

pamoroso commented:
Never mind, this PR has not been merged in master yet.

pamoroso commented:
The latest Medley Online build 26-Feb-2025 01:25:07 integrates this merged PR but pressing the arrow keys has no effect, perhaps because of noVNC. The arrow keys work fine on Linux with Medley VNC/TigerVNC.
